micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java

```diff
@@ -48,11 +53,12 @@
 public abstract class MeterRegistry {
     protected final Clock clock;
 
-    public MeterRegistry(Clock clock) {
+    protected MeterRegistry(Clock clock) {
         this.clock = clock;
     }
 
-    private final Map<Meter.Id, Meter> meterMap = new HashMap<>();
+    private final ConcurrentMap<Id, Meter> meterMap = new ConcurrentHashMap<>();
```

Comment:
This has caused deadlock in the past, will have to walk this back I think.

Reply:
It happens when you do nested registration in the builder function. Described at: https://bugs.openjdk.java.net/browse/JDK-8062841